torch/csrc/jit/codegen/cuda/arith.cpp

```diff
@@ -1684,22 +1688,113 @@ WelfordResult Welford(
   IrBuilder::create<WelfordOp>(
       out_avg,
       out_var,
-      out_N, /*out var/avg/count */
-      tv, /*in var/avg/count */
+      out_N, /*out avg/var/count */
+      tv, /*in avg/var/count */
       FusionGuard::getCurFusion()->zeroVal(),
       FusionGuard::getCurFusion()->oneVal(),
       init_avg_val,
       init_var_val,
-      init_N); /*init var/avg/count */
+      init_N); /*init avg/var/count */
 
   return WelfordResult(out_avg, out_var, out_N);
 }
 
+WelfordResult Welford(
+    TensorView* tv,
+    const std::vector<int>& axes,
+    TensorView* init_avg,
+    TensorView* init_var,
+    Int* init_N) {
+  TORCH_CHECK(
+      TensorDomain::sameAs(tv->getMaybeRFactorDomain(), tv->domain()->domain()),
+      "Reducing a tensor once it's gone under transformations is not permitted at this time. \n",
+      "Please set reductions before calling split/merge/computeAt.\n  RFactor: ",
+      tv->getMaybeRFactorDomain(),
+      "\n  Domain: ",
+      tv->domain()->toString());
+
+  TORCH_CHECK(tv->nDims() > 0, "Tried to reduce a 0-dim tensor");
+  TORCH_CHECK(axes.size() > 0, "No reduction axis specified");
+
+  // Check and collect reduction axes
+  std::vector<unsigned int> uint_axes;
+  auto tv_root = tv->domain()->noReductions();
+  const int ndims = tv_root.size();
+  for (int axis : axes) {
+    if (axis < 0) {
+      axis += ndims;
+    }
+
+    TORCH_CHECK(
+        axis >= 0 && axis < ndims,
+        "Reduction on invalid axis, received: ",
+        axis,
+        " however tensor view only has ",
+        ndims,
+        " non-reduction dims.");
+
+    uint_axes.push_back((unsigned int)axis);
+  }
+  std::sort(uint_axes.begin(), uint_axes.end());
+
+  // Squeeze before reduction
+  std::vector<int> reduction_axes;
+  std::vector<bool> is_trivial_reduction(ndims, false);
+  int offset = 0;
+  for (unsigned int axis : uint_axes) {
+    auto id = tv_root[axis];
+    is_trivial_reduction[axis] = id->isBroadcast() &&
+        !id->hasExpandedExtent() && id->extent()->isOneInt();
+    if (!is_trivial_reduction[axis]) {
+      reduction_axes.push_back(axis + offset);
+    } else {
+      offset--;
+    }
+  }
+
+  TensorView* squeezed = tv;
+  if (offset < 0) {
+    squeezed = squeeze(tv, is_trivial_reduction);
+  }
+
+  if (!reduction_axes.empty()) {
+    return WelfordRaw(squeezed, reduction_axes, init_avg, init_var, init_N);
+  }
+
+  // if squeeze only
+
+  if (init_N == nullptr) {
+    init_N = FusionGuard::getCurFusion()->zeroVal();
+  }
+  TensorView* out_N = full_like(squeezed, init_N, DataType::Index);
```

Comment:
So, when `init_N` is null (I don't know if we would ever need to support non-null `init_N`), `out_N` is zero. When all reductions are trivial, shouldn't the out N be `1`?

Reply:
I think what you said makes sense.

And I don't think non-null `init_N` is supported. Looks like non-zero init N is only useful to continue a welford from another welford, but this function only has one input.

Reply:
`WelfordOp` itself should support non-null `init_N`. When an initial `WelfordOp` is rfactored, the second `WelfrodOp` takes the output `N` from the first `WelfordOp`. But, here we're defining an interface for the user, and in that case I don't know `init_N` can be non-null. Anyway, that's a different topic.

Reply:
This is the code for rFactor of Welford:

```C++
    // Setup dependency chain, inserting producer before this op.
    // Expr* producer_definition =
    IrBuilder::create<WelfordOp>(
        producer_avg,
        producer_var,
        producer_n,
        wop->inAvg(),
        wop->inVar(),
        wop->inN(),
        wop->initAvg(),
        wop->initVar(),
        wop->initN());

    // Expr* consumer_definition =
    IrBuilder::create<WelfordOp>(
        wop->outAvg(),
        wop->outVar(),
        wop->outN(),
        producer_avg,
        producer_var,
        producer_n,
        wop->initAvg(),
        wop->initVar(),
        wop->initN());
```
Looks like we are using the same `initAvg`, `initVar` and `initN` for both `WelfordOp`. I don't know if this makes any sense. Shouldn't the second `WelfordOp` be totally ignoring these initial values?

Reply:
Sorry, my bad. The output `N` of the first `WelfordOp` should be the input `N` of the second `WelfordOp`. The init values in the second `WelfordOp` shouldn't be needed. What actually happens is the predicate of the second `WelfordOp` is just always `true` as the inputs are at least initialized, so the init values are never used in the runtime Welford functions.

Reply:
I have fixed this and added a new test `FusionSqueezeOnlyWelford_CUDA` for it